### PR TITLE
fix a scheduler panic due to internal cache inconsistency

### LINKED ISSUE #70450: kube-scheduler panic
**What happened**: kube-scheduler crashes (panics) in rare circumstances. Stack trace:

```
Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/cache/node_info.go:386
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:909
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:831
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:47
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x161223a]
goroutine 44674294 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x418ba40, 0x85fd7f0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
k8s.io/kubernetes/pkg/scheduler/cache.(*NodeInfo).Clone(0x0, 0xc428b44ac0)
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/cache/node_info.go:386 +0x3a
k8s.io/kubernetes/pkg/scheduler/core.selectVictimsOnNode(0xc42e66c700, 0x60fa9e0, 0xc422c60630, 0x0, 0xc4215a5e00, 0x61505a0, 0xc4201a7ab0, 0xc425744980, 0x5, 0x8, ...)
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:909 +0x7d
k8s.io/kubernetes/pkg/scheduler/core.selectNodesForPreemption.func1(0xf)
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:831 +0x176
k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.Parallelize.func1(0xc4300e3c50, 0xc4302c94a0, 0xc429e60380)
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:47 +0x96
created by k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.Parallelize
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:43 +0x10b
```

**What you expected to happen**: flawless operation ;)

**How to reproduce it (as minimally and precisely as possible)**: Unclear. The error above looks like it was associated with some cloning of node metadata. Perhaps due to a race during node scale-up/scale-down and cache interactions.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):

```
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:08:19Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```

/kind bug

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

When a node is deleted and genericScheduler.cachedNodeInfoMap isn't populated (called in Schedule()/Preempt()), invoking a nodeInfo.Clone() on a "stripped" NodeInfo will panic.

"stripped" means nodeInfo (the pointer) in schedulerCache side is taken off:

https://github.com/kubernetes/kubernetes/blob/e3ddaaad1d66ce1f74d6fa809d834c4de64546f9/pkg/scheduler/internal/cache/cache.go#L437

And most of its content are set to nil:

https://github.com/kubernetes/kubernetes/blob/843a67b215dcfa1781820a797c4c759e42e81d3a/pkg/scheduler/cache/node_info.go#L633

**Which issue(s) this PR fixes**:

Fixes #70450.

**Special notes for your reviewer**:

Generally speaking, the root cause is b/c internally we didn't handle the cache population properly. In schedulerCache, it watches on events of Node Add/Update/Delete, and update its nodeInfoMap immediately. But in a scheduling cycle, it just called `snapshot()` during `Schedule`()/`Preempt()`.

An ideal solution is to (1) do the cache "population" in real time, or (2) do you really need two nodeInfoMap? We can revisit it in our next refactoring iteration. cc/ @misterikkit .

This issue is likely to occur frequently in auto-scaler case - where node add/delete happens frequently.

**Does this PR introduce a user-facing change?**: 

```release-note
Fix a scheduler panic due to internal cache inconsistency
```

/sig scheduling
/assign @bsalamat

### CODE REVIEW

File: pkg/scheduler/core/generic_scheduler.go [hunk 1]
```diff
@@ -984,6 +983,10 @@ func selectVictimsOnNode(
 	queue internalqueue.SchedulingQueue,
 	pdbs []*policy.PodDisruptionBudget,
 ) ([]*v1.Pod, int, bool) {
+	if nodeInfo.Node() == nil {
```
Comment (justinsb) [inline pkg/scheduler/core/generic_scheduler.go]: Which line of Clone do we think is causing the panic here?  Looking at the nodeInfo.Clone() method body, I couldn't immediately spot which copy would panic on a stripped node, unless `nodeInfo` itself was nil.

Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler.go]: @justinsb from the stack trace, it's:

https://github.com/kubernetes/kubernetes/blob/076adbf4a2a4675efa040a6a03657d216755001a/pkg/scheduler/cache/node_info.go#L437

but honestly I can't find any code related with "setting requestedResource to nil" or "build NodeInfo() without init requestedResource".

And from the trace, it seems `nodeInfo` itself is not nil.

Comment (bsalamat) [reply pkg/scheduler/core/generic_scheduler.go]: @Huang-Wei I think the fix is to change `NodeInfo.Clone()` code to check for nil before calling Clone() on `requestedResource`, `nonzeroRequest`, and `allocatableResource`.

Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler.go]: Yes, we can do that - that's a more safe way.

My point on checking _if NodeInfo.node is nil_ was: if "nodeInfo.node == nil" is related with "nodeInfo.requestedResource == nil", (we haven't been able to confirm it yet), then we can directly return as calculation on a NodeInfoCopy with node==nil sounds error prone, or maybe we will see other panics.

As of now, I think we can make both changes on (1) checking "if NodeInfo.node is nil" and (2) if "requestedResource/nonzeroRequest/allocatableResource" is nil? Thoughts?

Comment (bsalamat) [reply pkg/scheduler/core/generic_scheduler.go]: I have no reason to believe that checking NodeInfo.node is nil helps with mitigating the crash. I would limit this to checking resources. Alternatively, we could dig to find where we initialize a NodeInfo object without initialing the "resources"

Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler.go]: ok, I will update to only check resources.

Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler.go]: Hmm, one interesting thing after reading the trace carefully, sharing with you: https://github.com/kubernetes/kubernetes/issues/70450#issuecomment-439253820

Comment (bsalamat) [reply pkg/scheduler/core/generic_scheduler.go]: Yes, I had seen that in the trace and that's why I suggested the fix above.

**Comment by justinsb** [comment]
So a question from trying to understand:

I don't see locking around cachedNodeInfoMap (and checkNode is spawned in a goroutine here, and I believe nodeNameToInfo is == cachedNodeInfoMap).

So is it possible that we concurrently deleted the node, and thus it is nil?

Technically we should be reading the map with a lock, because concurrent map access is unsafe.  However, given the likely perf impact and the slow rate of change here, my personal vote would be that (if I am correct that we should be using a lock) that we not address it till 1.14.  But we should check the case when the node is not in `nodeNameToInfo`?

(There are likely other ways also: we could also use a copy-on-write map, or maybe concurrent reading is safe if we are only writing from a thread that holds a write lock and we never change the set of keys etc)

File: pkg/scheduler/core/generic_scheduler.go (hunk 1 above)
Comment (ravisantoshgudimetla) [reply pkg/scheduler/core/generic_scheduler.go]: > Based on above comment, and that there are no chances requestedResource can be nil, I prefer to think the panic comes from a nil *NodeInfo.

I agree, I think it's the nodeInfo being nil causing the issue.

There are couple of ways to solve this, either at call site(place we are calling `nodeInfo.Clone()`, which are doing now in this PR) or within every method of nodeInfo where we can check if nodeInfo is nil. I don't have strong opinion as I see there are advantages of failing early(at the call site, which helps in friendly stack trace while debugging) going to method and checking it there(there is a good chance that someone might miss guarding against nil pointer when they are interacting with existing method assuming it is not going to panic, which happened in this case). 

If you want to do it at call site(like you are doing here), I'd recommend adding check for `nodeInfo==nil || nodeInfo.Node() == nil`

**Comment by Huang-Wei** [comment]
> So is it possible that we concurrently deleted the node, and thus it is nil?

This is not that possible. But it reminds me of another possibility that when a new node is added, it's possible cachedNodeInfoMap doesn't have that entry, yet. Why? Let me explain:

As I mentioned in description of this PR, schedulerCache is always up-to-date as it uses informers to watch changes on pods/nodes/etc. But the sync from schedulerCache to cachedNodeInfoMap is not realtime - it's called in snapshot(), by Schedule() and Preempt().

The issue occurs in Preempt(): 

- on L251, it called snapshot() to populate the nodeInfoMap, it's good so far and snapshot() is a bi-directional comparison, and it's safe

https://github.com/kubernetes/kubernetes/blob/d0c3cd182cffb3e722b14322fd1ca854a8bf62b0/pkg/scheduler/core/generic_scheduler.go#L251

- but on L259, it uses NodeLister to generate all available names, and it's up-to-date! In other words, in between L251 and L259, it's possible there is a new node is added to the cluster

https://github.com/kubernetes/kubernetes/blob/d0c3cd182cffb3e722b14322fd1ca854a8bf62b0/pkg/scheduler/core/generic_scheduler.go#L259

- and on L266, it calculates a potentialNodes, which is filter out nodes which has failed on previous run of scheduling, and for sure a new node won't be filtered out

https://github.com/kubernetes/kubernetes/blob/d0c3cd182cffb3e722b14322fd1ca854a8bf62b0/pkg/scheduler/core/generic_scheduler.go#L266

- then it comes to where the panic happens: calls selectNodesForPreemption() and in selectNodesForPreemption(), checkNode() runs to iterate on every potentialNodeName and call nodeNameToInfo[nodeName].Clone() directly:

https://github.com/kubernetes/kubernetes/blob/d0c3cd182cffb3e722b14322fd1ca854a8bf62b0/pkg/scheduler/core/generic_scheduler.go#L276

https://github.com/kubernetes/kubernetes/blob/d0c3cd182cffb3e722b14322fd1ca854a8bf62b0/pkg/scheduler/core/generic_scheduler.go#L904-L910

File: pkg/scheduler/core/generic_scheduler.go (hunk 1 above)
Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler.go]: > I'd recommend adding check for `nodeInfo==nil || nodeInfo.Node() == nil`

nodeInfo.Node() checks both if "nodeInfo == nil" or "nodeInfo.node == nil" :)

**Comment by AishSundar** [comment]
/milestone v1.13

Adding to 1.13. But do note that Code freeze is at 5pm today. If this is critical for 1.13 and need to get in, please switch priority to critical-urgent

**Comment by Huang-Wei** [comment]
I added a UT to simulate the panic:

```
Running tool: /usr/local/bin/go test -timeout 30s k8s.io/kubernetes/pkg/scheduler/core -run ^TestSelectNodesForPreemption$ -v -count=1

=== RUN   TestSelectNodesForPreemption
=== RUN   TestSelectNodesForPreemption/a_pod_that_does_not_fit_on_any_machine
=== RUN   TestSelectNodesForPreemption/a_pod_that_fits_with_no_preemption
=== RUN   TestSelectNodesForPreemption/a_pod_that_fits_on_one_machine_with_no_preemption
E1116 12:10:41.760498   60306 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/Users/wei.huang1/gospace/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/Users/wei.huang1/gospace/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/Users/wei.huang1/gospace/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/Cellar/go/1.11.1/libexec/src/runtime/asm_amd64.s:522
/usr/local/Cellar/go/1.11.1/libexec/src/runtime/panic.go:513
/usr/local/Cellar/go/1.11.1/libexec/src/runtime/panic.go:82
/usr/local/Cellar/go/1.11.1/libexec/src/runtime/signal_unix.go:390
/Users/wei.huang1/gospace/src/k8s.io/kubernetes/pkg/scheduler/cache/node_info.go:437
/Users/wei.huang1/gospace/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:990
/Users/wei.huang1/gospace/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:909
/Users/wei.huang1/gospace/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:65
/usr/local/Cellar/go/1.11.1/libexec/src/runtime/asm_amd64.s:1333
```

 and interestingly it's also pointing to the same line as original stack trace :)

```
		requestedResource:       n.requestedResource.Clone(),
```

**Review by ravisantoshgudimetla — APPROVED** [review]
Thanks @Huang-Wei for finding and fixing this bug.

/lgtm
/approve

/cc @bsalamat

File: pkg/scheduler/core/generic_scheduler.go (hunk 1 above)
Comment (bsalamat) [reply pkg/scheduler/core/generic_scheduler.go]: How does it check both? Besides, I am not sure if we should actually check nodeInfo.node == nil.

Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler.go]:
```go
func (n *NodeInfo) Node() *v1.Node {
	if n == nil {
		return nil
	}
	return n.node
}
```

nodeInfo.node == nil may not happen, as nodeInfo here is a cloned one, setting nodeInfo.node to nil can only happen in schedulerCache when it receive event of RemoveNode(). That one won't impact this cloned one.

So @bsalamat @ravisantoshgudimetla I can change it to check if nodeInfo == nil only, if it's more reasonable to you.

Comment (bsalamat) [reply pkg/scheduler/core/generic_scheduler.go]: I see your point Wei. You are right, but I still feel checking nodeInfo == nil is more readable for those who may be confused like myself.

Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler.go]: sure, will update.

**Comment by Huang-Wei** [comment]
@randomvariable that's what I'm going to do :)

File: pkg/scheduler/core/generic_scheduler.go (hunk 1 above)
Comment (ravisantoshgudimetla) [reply pkg/scheduler/core/generic_scheduler.go]: Yeah, at first I found it be confusing too but after looking at the code for a bit, I understood it.

**Comment by ravisantoshgudimetla** [comment]
@bsalamat any other comments before I re-lgtm?

**Review by bsalamat — COMMENTED** [review]
/lgtm
/approve

Thanks, @Huang-Wei and @ravisantoshgudimetla!

**Pull request merged**